Expose unifi PoE ports as individual switches

Flameeyes opened the pull request:
(Starting this in draft so that I can share it, but still need to be tested/form filled in/documentation updated.)

<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->


## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->


## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [x] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [x] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [x] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [x] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

Kane610 commented:
Thanks for PR! First, please split it, the switch part might warrant more discussions (and breaking changes) and to keep the size small, keep them apart per platform.

Flameeyes committed:
unifi: expose PoE ports as individual switches

Before this change, switches were provided only for devices that were at
least once discovered on a PoE-enabled port, but with the caveat that
only the last port and uplink would be preserved and restored to try
and turning them back on. This also restricted the availability of these
switches to clients, rather than UniFi devices themselves.

With this change, `SwitchEntity` are added per-port for every PoE port
that is found available, whether it has a client connected or not (since
there is no way for a device to know if there's a turned off client
connected to the port.) This allows to then label the switches
separately and re-enable whichever downlink is present, including other
unifi devices (e.g. to automate turning on or off additional access
points.)

Note that unlike the previous PoE switches that were only added if the
configuration requested it, these switches are added unconditionally.

Thanks to the ability to mark these as configuration switches, they are
not exposed in dashboards by default, and unlike the previous client PoE
switches, they are not unbounded: there's only at most as many as there
are ports on the switches themselves (most switches have only half of
their ports PoE-enabled.)

Flameeyes commented:
Split it to start with the switch. I'm also tracking a few more changes as a full branch on my repo for now, but will submit them individually so that any change discuss for this first one can follow up on the others.

Flameeyes commented:
@Kane610 I could use some historical context to figure out how to approach the config option. Right now the integration exposes the PoE controls only optionally — my guess would have been because it's a heuristic and not entirely reliable, and having a changing client on the port might just make it churn through a number of switch entities for no reason, but I also know that there's been discussions in the past of reducing the amount of data that the integration generates overall.

If we expose the ports as individual switches, the whole heuristics part is gone, and switch entities are no longer unbounded — if an user wants to attach a switch to a client, they can do so with a template switch on top of it. It does mean we'll be exposing a number of extra switches per device, which I personally think is okay, but it's only my opinion obviously.

Kane610 commented:
> @Kane610 I could use some historical context to figure out how to approach the config option. Right now the integration exposes the PoE controls only optionally — my guess would have been because it's a heuristic and not entirely reliable, and having a changing client on the port might just make it churn through a number of switch entities for no reason, but I also know that there's been discussions in the past of reducing the amount of data that the integration generates overall.
> 
> If we expose the ports as individual switches, the whole heuristics part is gone, and switch entities are no longer unbounded — if an user wants to attach a switch to a client, they can do so with a template switch on top of it. It does mean we'll be exposing a number of extra switches per device, which I personally think is okay, but it's only my opinion obviously.

IIRC (it was some time ago) there was no need for the switch port entities by themselves since they're typically only relevant in the context of a client (or device). What has been learned along the way is to not overdo it wrt providing too many entities, there have been requests at times to limit (or increase control of) how many entities (especially device trackers) that the integration creates. https://community.home-assistant.io/t/unifi-controller-integration-make-device-tracker-and-rx-tx-stat-sensors-optional-for-all-entities/341832

Certain aspects of HA has of course changed over the years, so it's not to say that the current implementation would be the best approach today. There are also more categories of entities today that help filter between them. 
With that said I'm not sure what we gain here except for a much simpler implementation.

I'll think about it more and get back to you shortly.

Flameeyes commented:
There's a couple of reasons why I went down this route:

 * Simplicity, as you noted, since the implementation doesn't have to guess or keep restoring the client switches while the client is not found. It also removes the confusion of a switch not appearing if you reuse the PoE port of a previous client (e.g. if you were to replace a PoE-powered device without changing port) or if the device connected happened to report multiple Mac addresses (e.g. if it has virtual machines on it.)
 * The ability to control ports even when there isn't something that's considered "a client" attached. This means that you can use a PoE splitter without an Ethernet device on the port, turning spare PoE ports into sort-of smart plugs.

To be honest this only makes sense now that the switches can be marked as configuration, as otherwise they would definitely be difficult to deal with in Cloud integrations and areas.

Kane610 commented:
I've been thinking about this, and it can be fine to add this solution. But I don't think the previous implementation should be removed, at least not initially as that will be a breaking change.

Some more things to consider is whether the user should select which ports to expose or if all should be exposed and if they do would be disabled by default. I'm personally leaning towards allowing the user to select what ports to expose, as that seems to be in what direction is typically requested.

jjlawren commented:
Reducing complexity is always good as long as no significant features are lost along the way.

I think controlling PoE per-port is intuitive and mirrors the concept as exposed by the Unifi controller interface. The convenience of being able to enable/disable based on the attached client is nice, but I'd expect the wired port a given client is attached to is relatively static and that abstraction probably isn't necessary.

Controlling PoE when there is not a detectable client associated with a port is a nice bonus.

jjlawren commented:
> Some more things to consider is whether the user should select which ports to expose or if all should be exposed and if they do would be disabled by default

I'd just say create a `switch` entity per-port (perhaps disabled by default) and associate them to the Unifi switch device, relying on the new `entity_category` as "config" to organize them appropriately. No config for the user besides enabling a `switch` entity for each port they want to control on the Device page.

The one thing that might make this tricky is that the `device_tracker` platform associates a device per-client with the Unifi integration. It might be hard to find the Unifi switch device. My setup has 1200+ "devices" associated with the Unifi integration. 😬

Flameeyes commented:
> I've been thinking about this, and it can be fine to add this solution. But I don't think the previous implementation should be removed, at least not initially as that will be a breaking change.

I was thinking of going down that route originally, but I got myself lost because then you have (to the bests of my attempts) there's two options:

 * create both the old and new switch, and let both control the same port;
 * only create the new switch if the old switch wouldn't be created.

In either case it adds a bit more complexity, so _personally_ I would prefer a "rip the bandaid" solution. But I'm happy to be told otherwise and I'll see to implement one or the other.

Also one note, if I understand the logic correctly (and I admit I haven't figured out how to check that for sure), right now you would find yourself in a situation where a port is turned off by Home Assistant and requires going to the controller to re-enabled:

 1. enable poe control of clients;
 2. connect a single client C1 to port P1 of a compatible switch; [HA creates a switch for C1]
 3. turn off C1 from Home Assistant;
 4. disconnect C1 from port P1 and connect it to P2; [HA updates C1 switch to point at P2, and consider the switch "on"]
 5. connect a new client C2 to port P1; [C2 stays turned off and can't be managed by HA]

(There's another case related to PoE mode, which is why in my current implementation it only goes auto/off/auto: if you had manual PoE mode selected, turn the switch off, then connect a _different_ client and turn it back on, you may not put it into a mode where the client can be powered on.)

> The one thing that might make this tricky is that the device_tracker platform associates a device per-client with the Unifi integration. It might be hard to find the Unifi switch device. My setup has 1200+ "devices" associated with the Unifi integration. 😬

Yeah the device tracker makes a few other things hard to find — but once you do, it's fairly straightforward:

![image](https://user-images.githubusercontent.com/74834/142483249-b474a243-0621-4f8c-b6ee-1ab8584a8f84.png)

This is how it looks with my stack of changes to the component: it includes the power/voltage per port (I renamed the port entities to find them more easily), and an update binary sensor.

jjlawren commented:
I think the model should be changed to simply expose the physical ports on the switch and to remove the concept of mapping connected clients for PoE entities. If a user wants to turn a given client on/off, they can rename the port via entity registry to represent that client.

Kane610 commented:
Ok, I've thought about it and we can go with the breaking change. I will need to re-review this thing :) I noticed stuff as referencing client when this is a more generic control. It should also clean up all existing POE client entities.

Flameeyes commented:
I've updated the pull request to no longer reference clients in the new switches. I've also avoided adding a new configuration option for this under the assumption that (unlike the previous implementation) this one is bounded by the ports available on the device. It makes sense for it to be exposed by default with that in mind — but I can be convinced otherwise.

As for cleaning up the existing POE client entities… do you have any pointer on how to do that? I'm afraid I'm not sure what to do besides leaving them as orphan after the update — this is incidentally the same that would happen if you turned the poe client setting off, AIUI.

(Note: both this and the follow-up changes I have use the new enums only for the newly added entities. if you want I can go ahead and do a first pass to update all the old ones onto them to make the other changes simpler.)

Kane610 commented:
I will try to review this soon. I'm all out of energy in the evenings right now

Kane610 reviewed (changes requested):
Overall it looks really good. Good job!

Is the test coverage kept 100%? Im also thinking a test to cover a new device with poe ports would be good as well.

You're typing some but not all, can you type the rest so a file is complete?

There is also a need to clean up the old PoE entities from the old implementation.

Kane610 commented on homeassistant/components/unifi/switch.py:
```diff
@@ -178,39 +122,20 @@ def add_dpi_entities(controller, async_add_entities, dpi_groups):
 
     _attr_entity_category = EntityCategory.CONFIG
+    _attr_device_class = SwitchDeviceClass.OUTLET
+    _attr_icon = "mdi:ethernet"
```
Is this really the best icon for PoE control?

stephanvierkant replied:
The common icon I've found is something like this: https://icons8.com/icon/25414/power-over-ethernet

So this might be the best option as long there isn't a dedicated PoE-icon.

Kane610 commented:
With how to clean up old entities:

Take inspiration from this one https://github.com/home-assistant/core/blob/943aaaeb3f1d1eabe4e0e46303f84ed539560236/homeassistant/components/unifi/unifi_entity_base.py#L80

and the logic to find all poe entities is here https://github.com/home-assistant/core/blob/943aaaeb3f1d1eabe4e0e46303f84ed539560236/homeassistant/components/unifi/switch.py#L58

Kane610 commented:
Line 338 of controller should be removed "or entry.unique_id.startswith(POE_SWITCH)" as there is no reason to restore clients for POE control when this PR is done

Kane610 commented:
Hey! There is still time to get this ready for 2022.2 release

stephanvierkant commented:
@Flameeyes Thanks for this PR! I'd really love to use this feature.

Are you still working on it? Is there any way we can help?

Flameeyes commented:
Hey folks, sorry, this ended up _a lot_ behind on my schedule :(

I'm going to try updating this to be compatible with 2022.4 tonight, but any help to get this complete is definitely appreciated!

Flameeyes reviewed (commented):
Updated to 2022.4.

Kane610 reviewed (changes requested):
Looks good so far! My previous comments still stand though.

HabelStefan commented:
How do I implement this in Hassio?

Kane610 commented:
> How do I implement this in Hassio?

You wait for the PR to be merged and then update core once it's been released.

roechas commented:
So to clarify, there still is not a simple integration with Unify to control the power of ports on the POE Switch?
But this is something in the works and will require a HASS core update?

gulli1986 commented:
Hello, is there any estimated time  for when this will be merged? Thanks!

Kane610 commented:
> Hello, is there any estimated time for when this will be merged? Thanks!

Author seems to have disappeared, Im doing some rewriting of the aiounifi library which will benefit this new approach so once Im further along with the rewrite I will revisit this PR and complete it.

Flameeyes commented:
> > Hello, is there any estimated time for when this will be merged? Thanks!
> 
> Author seems to have disappeared, Im doing some rewriting of the aiounifi library which will benefit this new approach so once Im further along with the rewrite I will revisit this PR and complete it.

I'm still around but haven't had time to commit more time to this unfortunately. The current code is "good enough" for my needs so I'd have to dedicate time to finish the PR.

If you'll be completing this I'll be happy to test!

Kane610 commented:
I can't give any promise to a time line but I want to proceed with the rewrite so shouldn't be too long

robloh commented:
I can also help test a rewrite

diogosalazar commented:
Are the failed checks significant above?

DohmenICT commented:
Hi! Any news on this? I really want to control a POE port on my Unifi switch.

I can't control it via the device MAC because this device (An amazon Fire tablet) is using wifi for a connection. So no MAC-address is detected hence I can't switch the POE off above battery 80% and switch the POE back on below 20%.

Kane610 commented:
> Are the failed checks significant above?

Yes

Kane610 commented:
> Hi! Any news on this? I really want to control a POE port on my Unifi switch.

Sorry, no set time line for this

Char-r commented:
@Kane610 I would be happy to sponsor the development of this feature if it would help? thanks

Kane610 commented:
> @Kane610 I would be happy to sponsor the development of this feature if it would help? thanks

Sponsoring is always appreciated! I'm doing the ground work of simplifying the integration which will also lead up to making this feature available for use.
